#! /bin/sh /usr/share/dpatch/dpatch-run
## 47_pid.dpatch by Kenshi Muto <kmuto@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Forwarded upstream: http://www.cups.org/str.php?L2465

@DPATCH@
diff -urNad cupsys-1.3.0~rc2~/scheduler/conf.c cupsys-1.3.0~rc2/scheduler/conf.c
--- cupsys-1.3.0~rc2~/scheduler/conf.c	2007-07-30 17:47:10.000000000 +0100
+++ cupsys-1.3.0~rc2/scheduler/conf.c	2007-08-01 13:23:12.000000000 +0100
@@ -167,7 +167,8 @@
 #endif /* HAVE_AUTHORIZATION_H */
   { "TempDir",			&TempDir,		CUPSD_VARTYPE_STRING },
   { "Timeout",			&Timeout,		CUPSD_VARTYPE_INTEGER },
-  { "UseNetworkDefault",	&UseNetworkDefault,	CUPSD_VARTYPE_BOOLEAN }
+  { "UseNetworkDefault",	&UseNetworkDefault,	CUPSD_VARTYPE_BOOLEAN },
+  { "PidFile",			&PidFile,		CUPSD_VARTYPE_STRING }
 };
 #define NUM_VARS	(sizeof(variables) / sizeof(variables[0]))
 
@@ -413,6 +414,7 @@
   cupsdSetString(&RemoteRoot, "remroot");
   cupsdSetString(&ServerHeader, "CUPS/1.2");
   cupsdSetString(&StateDir, CUPS_STATEDIR);
+  cupsdSetString(&PidFile, "/var/run/cups/cupsd.pid");
 #ifdef HAVE_GSSAPI
   cupsdSetString(&GSSServiceName, CUPS_DEFAULT_GSSSERVICENAME);
 #endif /* HAVE_GSSAPI */
diff -urNad cupsys-1.3.0~rc2~/scheduler/conf.h cupsys-1.3.0~rc2/scheduler/conf.h
--- cupsys-1.3.0~rc2~/scheduler/conf.h	2007-07-11 22:46:42.000000000 +0100
+++ cupsys-1.3.0~rc2/scheduler/conf.h	2007-08-01 13:19:41.000000000 +0100
@@ -186,6 +186,7 @@
 					/* Server key file */
 #  endif /* HAVE_LIBSSL || HAVE_GNUTLS */
 #endif /* HAVE_SSL */
+VAR char		*PidFile	VALUE(NULL); /* Debian CUPS pid file */
 
 #ifdef HAVE_LAUNCHD
 VAR int			LaunchdTimeout		VALUE(DEFAULT_TIMEOUT);
diff -urNad cupsys-1.3.0~rc2~/scheduler/main.c cupsys-1.3.0~rc2/scheduler/main.c
--- cupsys-1.3.0~rc2~/scheduler/main.c	2007-07-19 00:52:15.000000000 +0100
+++ cupsys-1.3.0~rc2/scheduler/main.c	2007-08-01 13:19:41.000000000 +0100
@@ -89,6 +89,8 @@
 static void		sigterm_handler(int sig);
 static long		select_timeout(int fds);
 static void		usage(int status);
+int			write_pid(void);
+int			remove_pid(void);
 
 
 /*
@@ -529,6 +531,11 @@
   }
 #endif /* __sgi */
 
+  if (write_pid() == 0) {
+    cupsdLogMessage(CUPSD_LOG_ERROR, "Unable to write pid file");
+    return (1);
+  }
+
  /*
   * Initialize authentication certificates...
   */
@@ -1009,9 +1016,39 @@
 
   cupsdStopSelect();
 
+  remove_pid();
+
   return (!stop_scheduler);
 }
 
+/* 'write_pid()' - Write PID file.
+   'remove_pid()' - Delete PID file.
+*/
+int
+write_pid()
+{
+  FILE *f;
+  int fd;
+  int pid;
+  if (((fd = open(PidFile, O_RDWR|O_CREAT, 0644)) == -1)
+      || ((f = fdopen(fd, "r+")) == NULL) ) {
+    return 0;
+  }
+  pid = getpid();
+  if (!fprintf(f, "%d\n", pid)) {
+    close(fd);
+    return 0;
+  }
+  fflush(f);
+  close(fd);
+
+  return pid;
+}
+
+int remove_pid() {
+  return unlink(PidFile);
+}
+
 
 /*
  * 'cupsdClosePipe()' - Close a pipe as necessary.
